pkg/controller/client.go

```diff
@@ -0,0 +1,194 @@
+	workloadStream *workload.WorkloadStream
+	xdsConfig      *config.XdsConfig
+}
+
+func NewXdsClient() *XdsClient {
+	client := &XdsClient{
+		xdsConfig: config.GetConfig(),
+		adsStream: &envoy.AdsStream{
+			Event: envoy.NewServiceEvent(),
+		},
+		workloadStream: &workload.WorkloadStream{
+			Event: workload.NewServiceEvent(),
+		},
+	}
+
+	client.ctx, client.cancel = context.WithCancel(context.Background())
+
+	return client
+}
+
+func (c *XdsClient) createStreamClient() error {
+	var err error
+
+	if c.grpcConn, err = nets.GrpcConnect(c.xdsConfig.DiscoveryAddress); err != nil {
+		return fmt.Errorf("grpc connect failed: %s", err)
+	}
+
+	c.client = service_discovery_v3.NewAggregatedDiscoveryServiceClient(c.grpcConn)
+
+	if bpfConfig.EnableKmesh {
+		if err = c.adsStream.AdsCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create ads stream failed, %s", err)
+		}
+	} else if bpfConfig.EnableKmeshWorkload {
+		if err = c.workloadStream.WorklaodCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create workload stream failed, %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *XdsClient) recoverConnection() error {
+	var (
+		err      error
+		interval = time.Second
+	)
+
+	c.closeStreamClient()
+	for count := 0; count < nets.MaxRetryCount; count++ {
+		if err = c.createStreamClient(); err == nil {
+			return nil
+		}
+
+		log.Errorf("grpc connect failed, %s", err)
+		c.closeStreamClient()
+		time.Sleep(interval + nets.CalculateRandTime(RandTimeSed))
+		interval = nets.CalculateInterval(interval)
+	}
+
+	return fmt.Errorf("retry %d times", nets.MaxRetryCount)
+}
+
+func (c *XdsClient) clientProcess(ctx context.Context) {
+	var (
+		err            error
+		reconnect      = false
+		isConnectAgain = new(bool)
+	)
+
+	*isConnectAgain = false
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if reconnect {
+				log.Warnf("reconnect due to %s", err)
+				if err = c.recoverConnection(); err != nil {
+					log.Errorf("recover connection failed, %s", err)
+					return
+				}
+				reconnect = false
+				if bpfConfig.EnableKmeshWorkload {
+					*isConnectAgain = true
+				}
+			}
+
+			if bpfConfig.EnableKmesh {
+				if err = c.adsStream.AdsStreamProcess(); err != nil {
+					reconnect = true
+					continue
+				}
+			} else if bpfConfig.EnableKmeshWorkload {
+				if err = c.workloadStream.WorkloadStreamProcess(isConnectAgain); err != nil {
+					reconnect = true
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (c *XdsClient) Run(stopCh <-chan struct{}) error {
+	if err := c.createStreamClient(); err != nil {
+		return fmt.Errorf("create stream failed, %s", err)
+	}
+
+	go c.clientProcess(c.ctx)
+
+	go func() {
+		<-stopCh
+
+		c.closeStreamClient()
+		if c.cancel != nil {
+			c.cancel()
+		}
+	}()
+
+	return nil
+}
+
+func (c *XdsClient) closeStreamClient() {
+	if bpfConfig.EnableKmesh {
+		if c.adsStream != nil && c.adsStream.Stream != nil {
+			_ = c.adsStream.Stream.CloseSend()
```

Comment:
Just don't handle the reture value if we don't need it. Only use "_" if there are multiple return values and we want to ignore some of them.